// sitefinity-webservices-sdk/config/webpack.common.js
// For instructions about this file refer to
// webpack and webpack-hot-middleware documentation
const webpack = require("webpack");
const helpers = require("./helpers");

module.exports = function(options) {
    return {
        mode: 'development',
        entry: {
            "app": helpers.appRoot("main")
        },

        output: {
            path: helpers.root("dist"),
            filename: "sitefinity-webservices-sdk.js",
            chunkFilename: "sitefinity-webservices-sdk.js",
            libraryTarget: "umd",
            library: "Sitefinity"
        },
        plugins: [
            // This plugin optimizes chunks and modules by how much they are used in your app
            new webpack.optimize.OccurrenceOrderPlugin(),
            //Skipping the error emitting phase (and recording phase) during compilation, so there are no assets emitted that include errors
            new webpack.NoEmitOnErrorsPlugin()
        ],

        resolve: {
            extensions: [".ts", ".js", ".json", ".html"]
        },

        module: {
            rules: [
                {
                    test: /\.ts$/,
                    use: ["awesome-typescript-loader"]
                }
            ]
        }
    };
};
